kntest/pkg/cluster/options.go

```diff
@@ -0,0 +1,41 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"github.com/spf13/cobra"
+
+	"knative.dev/test-infra/kntest/pkg/cluster/ops"
+)
+
+func addOptions(clusterCmd *cobra.Command, rw *ops.RequestWrapper) {
+	pf := clusterCmd.PersistentFlags()
+	req := &rw.Request
+	pf.Int64Var(&req.MinNodes, "min-nodes", 0, "minimal number of nodes")
+	pf.Int64Var(&req.MaxNodes, "max-nodes", 0, "maximal number of nodes")
+	pf.StringVar(&req.NodeType, "node-type", "", "node type")
```

Comment:
The rest of these should not be arguments -- they need to be passed in via the environment.

Think about how Redhat folks would update this to work in their environment

Reply:
As we discussed in  https://github.com/knative/test-infra/pull/1883#discussion_r402314355, I'm going to defer the changes for env vars to following PRs.

And one more note: I believe env vars can sometimes cause negative effects, as you cannot restrict on where they can be set. That said, they can be set literally anywhere that makes them hard to manage and discover, that's exactly one of the problems we have in the other project. I think another option we can go with would be always using flags, and we can define groups of flags for each subcommand, e.g. for cluster related command, we can pass the flags like `--cluster "--project=xxx --name=xxx --version=xxx"`, in this way all the settings will be managed in one place, mandatorily.

This is of course just my thoughts, and we can certainly have more discussions. Also loop @chaodaiG @yt3liu in for more ideas.

Reply:
Added one more comment above. Agree that command line arguments are more preferable to me.

Reply:
I'd be okay if all cluster creation args were passed as one variable, like you wrote.

Reply:
Except what if one flow needs to say "make a cluster twice as big as normal"... how would we do that with this regime? (maybe we don't need to support that)
It wouldn't be any easier with env vars though.
Config file could be copied and do *2 on an arg.

Reply:
Just to be clear, the place I will make this change is at https://github.com/knative/test-infra/blob/master/scripts/e2e-tests.sh#L462-L471, and I want to do it in following PRs as this one is big enough.

For the example you gave, please correct me if I got it wrong, but I think it's not adding much trouble as users can do *2 before passing them as flags?

Reply:
We have one function to create clusters in the scripts. Currently it only works with GKE; not KIND or anything else.

So each e2e flow must have the exact same cluster creation.

But, for example, the autotls test flow shouldn't need the same size cluster, right? But it does run with the same one.

So if are passing in all arguments as one group of args (like `--cluster "--project=xxx --name=xxx --version=xxx"`) then how would we make the autotls test use half as many nodes?

What if we want a test flow with and without autoscaling?

None of this is being made easier with arguments; if anything, it's now harder than it was before.

Reply:
Even with the current test setup, we do not support disabling autoscaling for the GKE cluster. If we do need it in the future, we can add it as an option in the `create` command.

The `create` command now supports `--min-nodes` and `--max-nodes` option, developers can set them if they want a different size cluster. "half as many nodes" is a pretty vague request and I doubt people would want that...

Reply:
If we accept that flows are never going to have any input into the cluster created for them (never asking for addons or anything else) --- to say another way, if the only configuration allowed is in Prow or on the command line before running the scripts, then `--cluster-args="--project=xxx -name=xxx` will suffice.

Reply:
This discussion is going too far away from the original topic. I agree that it's something we need to solve but it's out of the scope of this PR.
We can certainly discuss the different options, including env vars, multiple args and one single arg, and choose one to move forward, but it should not be a blocker for this PR.

Reply:
I disagree. If the shell scripts aren't broken right now, there's no rush. We need to get this right. There's no sense in putting bad code in.
If there's no scenario where scripts need to modify args to kntest (other than possibly add some), then this approach is fine.

Reply:
Then what do you want? Env vars?

Reply:
If you can't think of a scenario where we need scripts to modify args to kntest, then lets move forward.